python/BioSimSpace/Sandpit/Exscientia/_SireWrappers/_molecule.py

```diff
@@ -171,7 +171,7 @@ def __contains__(self, other):
                             "or 'BioSimSpace._SireWrappers.Residue'.")
 
         # Return whether the object comes from this molecule.
-        return self._sire_object.molecule() == other._sire_object.molecule()
+        return self._sire_object.molecule().number() == other._sire_object.molecule().number()
```

Comment:
Is it worth overriding `__eq__` of the wrapped objects?

Reply:
@msuruzhon Would that be a Sire issue instead of a BSS issue?

Reply:
Sorry I meant that `__eq__` of the `_SireWrappers.Molecule` should wrap this equals statement. Or is that not desirable in general?

Reply:
@msuruzhon I guess the core issue is that 

`BSS.molecule._sire_object == BSS.molecule._sire_object`

With regard to whether it is desirable, I just only @lohedges could comment.

Reply:
In this case `other` could be an `Atom`, `Molecule`, or `Residue`, so this isn't a comparison at the molecule level. Calling `.molecule()` on a Sire object gets back the molecule to which the object belongs, but the comparison between those molecules doesn't appear to always be consistent. (For reference, you can already compare BioSimSpace molecules.)

```python
protein = BSS.IO.readMolecules(['/test/input/molecules/1jr5.crd', '../test/input/molecules/1jr5.top']).getMolecule(0)
ligand = BSS.IO.readMolecules(BSS.IO.glob("test/input/ligands/ligand01*")).getMolecule(0)

system = (ligand + protein).toSystem()

protein._sire_object.molecule() == system[1]._sire_object.molecule()
False
protein._sire_object.molecule().number() == system[1]._sire_object.molecule().number()
True

protein._sire_object
Molecule( 633 version 21 : nAtoms() = 1495, nResidues() = 90 )
system[1]._sire_object
Molecule( 633 version 1606 : nAtoms() = 1495, nResidues() = 90 )
```

I assume that the comparison is failing since the version numbers differ.

Reply:
@lohedges Make sense

Reply:
I'm just trying to work out at which point the version number changes and why, and if it's important.

Reply:
Yes - that's correct. Sire tests equality by comparing the ID of the object and the version number. This enables us to do quick tests for equality without having to run through and compare all of the data in the molecule.

A change in version number implies that `.edit().set(...).commit()` has been called on the molecule somewhere. This could be either because the molecule has been translated, or it has its properties changed etc.

In general, there are there levels of comparison you can do:

1. Just compare the ID numbers of the objects (e.g. molecule.number() == other.number()). This only guarantees that the molecules were the same at one point in the past (the molecule number does not change).
2. Compare the UID of the molecule.data().info() object (the MoleculeInfo). This is the layout of the molecule, i.e. names of atoms, arrangement into residues etc. Every layout that is the same has the same UID. Having the same molecule number and layout UID means that the molecule is the same and it has the same atoms, residues, etc
3. Compare the version number of the object. Every time the object changes the version number will increase. Two objects with the same ID number and version number are guaranteed to be exactly the same.

This is why a general ".contains" is difficult to implement. It is ambiguous, as do you mean "this molecule contains an atom with this name", or "this version of this molecule contains the same version of the atom".

Reply:
Yes, I agree with what @chryswoods says. Perhaps it's not worth including this feature yet because of the ambiguity, or at least document the operator so the use case is clear.

Reply:
I think one way to go would be
`is` will check both version number and MolNum
'__eq__' will only check MolNum?

Reply:
Currently it's the opposite, which makes more sense to me, despite still being ambiguous. As mentioned above, this is because, when creating a system, atoms and residues are renumbered to be unique and in ascending order. This results in the following situation:

```python
s = BSS.IO.readMolecules("amber/ala/*")
ss = (s[1] + s[0]).toSystem()

# Check by number only.
s[0] in ss
True

# Check by number and version.
s[0] == ss[1]
False
```

It might be possible to move the renumbering code into `Process.Somd`, so that this is done once on setting things up for a SOMD simulation. I'll need to check that this doesn't cause other issues, though.

Reply:
`__eq__` implies equality. You could, in theory, change both the layout and the properties of a molecule while retaining the same molecule number. This would, in the worst case, mean you could have `__eq__` give true for two molecules that have different numbers of atoms. I would prefer something like

```
atom.molecule().number() in system
```

(i.e. you are explicitly asking if this molecule number is in a system)

I remember that Lester mentioned that he had a general "ID", so it could be

```
atom.ID() in system
```

Reply:
@lohedges Sorry, I mean `is` not `in`. The `is` could have a different behaviour to '__eq__'

Reply:
I see what you are saying, and it would make sense if this `__eq__` and `is` was used consistently throughout the rest of the code. However, because C++ doesn't have an `is`, it would be really hard to keep this use consistent. It would also be a potential source of surprise for people who didn't know about `is`. In general, I prefer `__eq__` to mean "these two things are identical - they are equal to each other". You can have two objects held in different bits of memory that have the same data, so would be `__eq__` but would not be `is` (`is` means that the two variables point to the same piece of memory).

I know it is a little annoying. But these different levels of comparison exist because there are different types of equality questions that people could ask, i.e. "these are exactly the same molecules", "these molecules have exactly the same data", "these two molecules have the same internal layout" etc.

Following "explicit is better than implicit", I prefer using `object in container` to mean the exact same data of object is in the container. Then `object.ID() in container` to mean that an object with the same ID is in the container. And then, `object.layout() in container` to mean that the container has an object with the same internal layout. (note that all three of these are pseudocode - we don't have a `object.layout()` function, although it may be a good thing to add)

Reply:
I think I'll just try moving this bit...

```python
self._sire_object = _SireIO.renumberConstituents(self._sire_object, num_mols)
```

into `setup` method of `Process.Somd`. This way both `in` and `eq` will behave in _exactly_ the same way, i.e. checking for a molecule with the same number and version.

I'll just need to make sure that there aren't any unintended consequences of doing this, i.e. any other things that assume unique numbering elsewhere.

Reply:
This is actually more complicated that I thought, and will require and update to the C++ wrapper that does the renumbering.

Reply:
I've added some updates to Sire [here](https://github.com/michellab/Sire/tree/fix_bss_333) that resolve the issue with moving the renumbering functionality into the SOMD setup. (Previously the coordinate/velocity update wrapper assumed that the system had been renumbered, where now we leave the original system numbering untouched.) This seems to work, and allows us the `__contains__` operator to work on molecules, not molecule numbers.

I'll still have a think about how we want to deal with ambiguity, i.e. by providing the option have different levels of comparison, as suggested by @chryswoods. (Check only by number, or by number and version, etc.)